// nuxi/dist/chunks/upgrade.mjs
import { execSync } from 'child_process';
import { promises } from 'fs';
import { d as defineNuxtCommand, c as consola } from './index.mjs';
import { r as resolve } from './index2.mjs';
import { r as resolveModule } from './cjs.mjs';
import { g as getPackageManager, p as packageManagerLocks } from './packageManagers.mjs';
import 'tty';
import 'util';
import 'path';
import 'os';
import 'assert';
import 'module';
import 'url';
import './fs.mjs';
import './rimraf.mjs';

async function getNuxtVersion(paths) {
  const pkgJson = resolveModule("nuxt3/package.json", paths);
  const pkg = pkgJson && JSON.parse(await promises.readFile(pkgJson, "utf8"));
  if (!pkg.version) {
    consola.warn("Cannot find any installed nuxt3 versions in ", paths);
  }
  return pkg.version || "0.0.0";
}
const upgrade = defineNuxtCommand({
  meta: {
    name: "upgrade",
    usage: "npx nuxi upgrade [--force|-f]",
    description: "Upgrade nuxt3"
  },
  async invoke(args) {
    const rootDir = resolve(args._[0] || ".");
    const packageManager = getPackageManager(rootDir);
    if (!packageManager) {
      console.error("Cannot detect Package Manager in", rootDir);
      process.exit(1);
    }
    const packageManagerVersion = execSync(`${packageManager} --version`).toString("utf8").trim();
    consola.info("Package Manager:", packageManager, packageManagerVersion);
    const currentVersion = await getNuxtVersion(rootDir);
    consola.info("Current nuxt version:", currentVersion);
    if (args.force || args.f) {
      consola.info("Removing lock-file and node_modules...");
      await Promise.all([
        promises.rm(packageManagerLocks[packageManager]),
        promises.rmdir("node_modules", { recursive: true })
      ]);
      execSync(`${packageManager} install`, { stdio: "inherit" });
    } else {
      consola.info("Upgrading nuxt...");
      execSync(`${packageManager} ${packageManager === "yarn" ? "add" : "install"} -D nuxt3@latest`, { stdio: "inherit" });
    }
    const upgradedVersion = await getNuxtVersion(rootDir);
    consola.info("Upgraded nuxt version:", upgradedVersion);
    if (upgradedVersion === currentVersion) {
      consola.success("You're already using the latest version of nuxt3.");
    } else {
      consola.success("Successfully upgraded nuxt from", currentVersion, "to", upgradedVersion);
      const commitA = currentVersion.split(".").pop();
      const commitB = upgradedVersion.split(".").pop();
      if (commitA && commitB) {
        consola.info("Changelog:", `https://github.com/nuxt/framework/compare/${commitA}...${commitB}`);
      }
    }
  }
});

export { upgrade as default };
